R/convert_output_types.R

```diff
@@ -0,0 +1,251 @@
+#'    specify when both `"quantile"` and `"cdf"` are desired)
+#' @param n_samples `numeric` that specifies the number of samples to use when
+#'   calculating output_types from an estimated quantile function. Defaults to `1e4`.
+#' @param ... parameters that are passed to `distfromq::make_q_fn`, specifying
+#'   details of how to estimate a quantile function from provided quantile levels
+#'   and quantile values for `"quantile"` or `"cdf"` output types.
+#'
+#' @details
+#' The following transformations are supported: (i) `"sample"` can be
+#' transformed to `"mean"`, `"median"`, `"quantile"`, or `"cdf"`; (ii)
+#' `"quantile"` can be transformed to `"mean"`, `"median"`, or `"cdf"`; and
+#' (iii) `"cdf"` can be transformed to `"mean"`, `"median"`, or `"quantile"`.
+#'
+#' For `"quantile"` and `"cdf"` starting output types, we use the following approach:
+#'   1. Interpolate and extrapolate from the provided quantiles or probabilities
+#'        for each component model to obtain an estimate of the cdf of that distribution.
+#'   2. Draw samples from the distribution for each component model. To reduce
+#'        Monte Carlo variability, we use quasi-random samples corresponding to
+#'        quantiles of the estimated distribution.
+#'   3. Calculate the desired quantity (e.g., mean).
+#' If the median quantile is provided in the `model_out_tbl` object (i.e., the
+#' original output_type is `"median"` and 0.5 is contained in the original
+#' output_type_id), the median value is extracted and returned directly.
+#'
+#' If both `"quantile"` and `"cdf"` output_types are desired, `new_output_type_id`
+#' should be a named list, where each element specifies the corresponding
+#' `new_output_type_id`. See examples for an illustration.
+#'
+#' @examples
+#' # We illustrate the conversion between output types using normal distributions,
+#' ex_quantiles <- c(0.25, 0.5, 0.75)
+#' model_out_tbl <- expand.grid(
+#'   group1 = c(1,2),
+#'   model_id = "A",
+#'   output_type = "quantile",
+#'   output_type_id = ex_quantiles
+#' ) %>%
+#' dplyr::mutate(value = qnorm(p = output_type_id, mean = group1))
+#'
+#' convert_output_type(model_out_tbl, c("group1"), new_output_type = "median")
+#'
+#' # Next, we illustrate conversion from samples to quantile and cdf
+#' ex_bins <- seq(-2,2,1)
+#' ex_quantiles <- c(0.25, 0.5, 0.75)
+#' model_out_tbl <- expand.grid(
+#'   group1 = c(1,2),
+#'   model_id = "A
+#'   output_type = "sample",
+#'   output_type_id = 1:100
+#' ) %>%
+#' dplyr::mutate(value = rnorm(100, mean = group1))
+#'
+#' convert_output_type(model_out_tbl, new_output_type = c("quantile", "cdf"),
+#'    new_output_type_id = list("quantile" = ex_quantiles, "cdf" = ex_bins))
+#'
+#' @return object of class `model_out_tbl` containing new output_type
+#' @export
+convert_output_type <- function(model_out_tbl, new_output_type,
+                                new_output_type_id = NA, n_samples = 1e4, ...) {
```

Comment:
The logic of this function, where everything is converted to samples and from there to the target output type, has a certain kind of elegance and at the same time is kind of inefficient.  For example, here's what we do when the `starting_output_type` is "quantile" and the `new_output_type` is "cdf":
1. In `get_samples_from_quantiles`, we call `distfromq::make_q_fn` to estimate the quantile function and draw samples from it
    - internally, to get an estimated quantile function `distfromq::make_q_fn` [estimates the cdf and inverts that](https://reichlab.io/distfromq/reference/make_q_fn.html#details)
2. Then we compute the empirical cdf of those samples and evaluate that ecdf in 

Restating, our goal was to get to cdf values, and in order to do that we (1) estimate the cdf; (2) invert that to get to a quantile function estimate; (3) draw samples from the quantile function; (4) get the empirical cdf of those samples; (5) evaluate that empirical cdf.  Although it is conceptually clean to work through the "universal representation" of samples, steps 2-5 here are extra work and in particular the sample-based stuff in steps 3-5 will introduce more layers of approximation to the result.

Another alternative organization to this code might provide a set of "direct" transformations that don't use samples as an intermediate step.  For example, the quantile to cdf transform might define a helper function along the lines of `transform_quantile_to_cdf_one_group` along these lines:
```
transform_quantile_to_cdf_one_group <- function(starting_ps, starting_qs, to_qs, ...) {
  p_fn <- distfromq::make_p_fn(
    ps = as.numeric(starting_ps),
    qs = starting_qs,
    ...
  )
  return(p_fn(to_qs))
}
```

Which could then be called on a grouped data frame somewhere, like
```
  samples <- model_out_tbl %>%
    dplyr::group_by(model_id, dplyr::across(dplyr::all_of(task_id_cols))) %>%
    dplyr::reframe(
      value = transform_quantile_to_cdf_one_group(
        starting_ps=as.numeric(.data$output_type_id),
        starting_qs=.data$value,
        to_qs=new_output_type_id,
        ...
      )
```

I think this approach would likely involve more code in the end, but could be nice as it would be more direct and involve less approximation.  I'm open to your ideas about this!

Reply:
Originally, I had conceptualized a set of transformations like `transform_quantile_to_cdf_one_group()` that you describe @elray1. When most required transforming to samples, I decided for the elegant route. But I agree, there is additional approximation that is not necessary, and would ideally be avoided. 

Question if we opt for the individual transformation functions: consider, for example, we are transforming from quantile to mean. The steps are (1) estimate the cdf, (2) invert to quantile function, (3) draw samples from quantile function, (4) calculate mean. Would this function call the `transform_quantile_to_samples()` function? Or would it replicate steps 1-3 within the `transform_quantile_to mean()` function? 

@lshandross, I'm happy for you to make the final decision on this one.

Reply:
This comment was a good prompt for me to think about which transformations require samples, estimates of the QF, and/or estimates of the CDF.  Here's my thinking, does this seem corret?

- starting type = quantile
    - new type = quantile: do we support this? should we? if so, would not require samples, use estimated QF
    - new type = cdf: samples not needed, use estimated CDF
    - new type = pmf: samples not needed, use estimated CDF.  Some data structure would be required specifying bin endpoints, or some convention for the names of the `output_type_id`s for bins, e.g. `(0,0.1]`.
    - new type = sample: **samples needed**
    - new type = median: samples not needed, use estimated QF
    - new type = mean: **samples needed**
- starting type = cdf
    - new type = quantile: samples not needed, use estimated QF
    - new type = cdf: do we support this? should we? if so, would not require samples, use estimated CDF
    - new type = pmf: samples not needed, use estimated CDF
    - new type = sample: **samples needed**
    - new type = median: samples not needed, use estimated QF
    - new type = mean: **samples needed**
- starting type = pmf.
    - overall note, i think that for the most part, transforms here only make sense if the variable is a discretization of an underlying continuous variable and we get some information about the bin endpoints used for the discretization.
    - though if the variable is nominal, we could go to the CDF type naturally.
- starting type = median
    - can't do any transformations other than the trivial transform to quantile at level 0.5
- starting type = mean
    - can't do any transformations

Reply:
Discussed this with Li just now, and arrived at a plan to merge this PR more or less as-is, then do the following in separate issues:

- Allow new_type = sample for existing probabilistic starting types where they are not supported (quantile and cdf)
- Allow new_type = pmf for existing probabilistic starting types (quantile, cdf, and sample).
    - Maybe at the same time, move away from using samples as intermediary for new type = cdf.  This would get us to a place where the same methodology is used when the new output type is pmf and cdf.
    - Maybe at the same time, support cdf -> cdf, e.g. extending the set of points where the cdf is evaluated.  This feels like a low priority extension, not sure there is really a use case?
- Allow `starting_type` = pmf
- Lower priority issue, support quantile --> quantile transformations, e.g. extending the set of quantile levels predicted. This could be useful to modelers who are doing something like quantile regression, to allow them to get predictions at a small number of quantile levels and then fill in the others.
    - At this time, could switch to using QF as an intermediary when new_type = quantile or median rather than samples.

Reply:
This sounds great, thank you both for pushing this across the finish line!